src/error-base.js

```diff
@@ -0,0 +1,526 @@
+    const messagingOrigin = self.AMP.viewer.maybeGetMessagingOrigin();
+    if (resolvedViewerUrl) {
+      data['rvu'] = resolvedViewerUrl;
+    }
+    if (messagingOrigin) {
+      data['mso'] = messagingOrigin;
+    }
+  }
+
+  if (!detectedJsEngine) {
+    detectedJsEngine = detectJsEngineFromStack();
+  }
+  data['jse'] = detectedJsEngine;
+
+  const exps = [];
+  const experiments = experimentTogglesOrNull(self);
+  for (const exp in experiments) {
+    const on = experiments[exp];
+    exps.push(`${exp}=${on ? '1' : '0'}`);
+  }
+  data['exps'] = exps.join(',');
+
+  if (error) {
+    const tagName = error.associatedElement
+      ? error.associatedElement.tagName
+      : 'u'; // Unknown
+    data['el'] = tagName;
+
+    if (error.args) {
+      data['args'] = JSON.stringify(error.args);
+    }
+
+    if (!isUserError && !error.ignoreStack && error.stack) {
+      data['s'] = error.stack;
+    }
+
+    // TODO(jridgewell, #18574); Make sure error is always an object.
+    if (error.message) {
+      error.message += ' _reported_';
+    }
+  } else {
+    data['f'] = filename || '';
+    data['l'] = line || '';
+    data['c'] = col || '';
+  }
+  data['r'] = self.document.referrer;
+  data['ae'] = accumulatedErrorMessages.join(',');
+  data['fr'] = self.location.originalHash || self.location.hash;
+
+  pushLimit(accumulatedErrorMessages, message, 25);
+
+  return data;
+}
+
+/**
+ * Resets accumulated error messages for testing
+ */
+export function resetAccumulatedErrorMessagesForTesting() {
+  accumulatedErrorMessages = [];
+}
+
+/**
+ * Does a series of checks on the stack of an thrown error to determine the
+ * JS engine that is currently running. This gives a bit more information than
+ * just the UserAgent, since browsers often allow overriding it to "emulate"
+ * mobile.
+ * @return {string}
+ * @visibleForTesting
+ */
+export function detectJsEngineFromStack() {
+  /** @constructor */
+  function Fn() {}
+  Fn.prototype.t = function() {
+    throw new Error('message');
+  };
+  const object = new Fn();
+  try {
+    object.t();
+  } catch (e) {
+    const {stack} = e;
+
+    // Safari only mentions the method name.
+    if (startsWith(stack, 't@')) {
+      return 'Safari';
+    }
+
+    // Firefox mentions "prototype".
+    if (stack.indexOf('.prototype.t@') > -1) {
+      return 'Firefox';
+    }
+
+    // IE looks like Chrome, but includes a context for the base* stack line.
+    // Explicitly, we're looking for something like:
+    // "    at Global code (https://example.com/app.js:1:200)" or
+    // "    at Anonymous function (https://example.com/app.js:1:200)"
+    // vs Chrome which has:
+    // "    at https://example.com/app.js:1:200"
+    // * The stack trace is limited to 10 lines, so we aren't really looking at
+    //   the base stack line, but it is a deep one, nonetheless.
+    const splitStack = stack.split('\n');
+    const last = splitStack[splitStack.length - 2];
```

Comment:
This change is a little iffy. The default stack trace from node is only 10 lines long, so we're likely not looking at the "base" stack line, as the original comment mentions. Moving `detectJsEngineFromStack` to `error-base` shifts the stack and the original line that was analyzed is now gone, invalidating this test (see examples below). This change works in one single test, but more evidence is needed to determine if it is reliable.

When `detectJsEngineFromStack` method exists in `error-base.js` and is invoked by `test/unit/test-error.js`:

    Error: message
        at Fn.t (http://localhost:9876/home/mdmower/source/amphtml/src/error-base.js:487:11)
        at detectJsEngineFromStack (http://localhost:9876/home/mdmower/source/amphtml/src/error-base.js:491:12)
        at detectJsEngineFromStack (http://localhost:9876/home/mdmower/source/amphtml/src/error.js:371:10)
        at Context.<anonymous> (http://localhost:9876/home/mdmower/source/amphtml/test/unit/test-error.js:679:14)
        at Context.<anonymous> (http://localhost:9876/home/mdmower/source/amphtml/test/_init_tests.js:250:17)
        at callFn (node_modules/mocha/mocha.js:5342:21)
        at Test.Runnable.run (node_modules/mocha/mocha.js:5334:7)
        at Runner.runTest (node_modules/mocha/mocha.js:5990:10)
        at node_modules/mocha/mocha.js:6108:12
        at next (node_modules/mocha/mocha.js:5902:14)

When `detectJsEngineFromStack` method exists in `error.js` and is invoked by `test/unit/test-error.js`:

    Error: message
        at Fn.t (http://localhost:9876/home/mdmower/source/amphtml/src/error.js:374:11)
        at detectJsEngineFromStack (http://localhost:9876/home/mdmower/source/amphtml/src/error.js:378:12)
        at Context.<anonymous> (http://localhost:9876/home/mdmower/source/amphtml/test/unit/test-error.js:679:14)
        at Context.<anonymous> (http://localhost:9876/home/mdmower/source/amphtml/test/_init_tests.js:250:17)
        at callFn (node_modules/mocha/mocha.js:5342:21)
        at Test.Runnable.run (node_modules/mocha/mocha.js:5334:7)
        at Runner.runTest (node_modules/mocha/mocha.js:5990:10)
        at node_modules/mocha/mocha.js:6108:12
        at next (node_modules/mocha/mocha.js:5902:14)
        at node_modules/mocha/mocha.js:5912:7